f2/src/utils/scanFiles.js

```diff
@@ -76,34 +78,59 @@ const contentModerateFile = (file, callback) => {
         message: 'Error in Content Moderator',
         error: err,
       })
-      file[1].adultClassificationScore = 'Could not scan'
+      file.adultClassificationScore = 'Could not scan'
     } else {
       try {
         const contMod = JSON.parse(response.body)
-        file[1].isImageRacyClassified = contMod.IsImageRacyClassified
-        file[1].isImageAdultClassified = contMod.IsImageAdultClassified
-        file[1].adultClassificationScore = contMod.AdultClassificationScore
-        file[1].racyClassificationScore = contMod.RacyClassificationScore
+        file.isImageRacyClassified = contMod.IsImageRacyClassified
+        file.isImageAdultClassified = contMod.IsImageAdultClassified
+        file.adultClassificationScore = contMod.AdultClassificationScore
+        file.racyClassificationScore = contMod.RacyClassificationScore
       } catch (error) {
         console.warn(`Error in Content Moderator: ${error.stack} `)
       }
     }
-    callback(null, file[1])
+    callback(null, file)
   })
 }
 
+//Queue to call content moderator once per second.
+const queue = async.queue((task, callback) => {
+  try {
+    /*
+      If time since last request is greater than 1 second, make the request immmediately.
+      Otherwise, wait until 1 second has passed to make the request.
+    */
+    const currentTime = new Date().getTime()
+    const timeDif = currentTime - lastRequest
+    const timeToWait = timeDif > 1000 ? 0 : 1000
+
+    setTimeout(() => {
+      lastRequest = new Date().getTime()
+      contentModerateFile(task, (err, _results) => {
+        if (err) {
+          console.warn('Content Moderator Error:' + JSON.stringify(err))
+        }
+      })
+      callback()
+    }, timeToWait)
+  } catch (err) {
+    console.warn('Error calling Content Moderator:' + JSON.stringify(err))
+    callback()
+  }
+}, 1)
+
 async function contentModeratorFiles(data, finalCallback) {
-  if (!serviceKey)
+  if (!serviceKey) {
     console.warn('Warning: files not scanned with Content Moderator')
-  else {
-    async.map(
-      Object.entries(data.evidence.files),
-      contentModerateFile,
-      function (err, _results) {
-        if (err) console.warn('Content Moderator Error:' + JSON.stringify(err))
-        finalCallback()
-      },
-    )
+  } else {
+    //Add files to queue
+    queue.push(data.evidence.files)
+
+    await queue.drain()
```

Comment:
Would this cause everybody to wait for everybody?

Reply:
This will cause all requests to be processed before submission. My thinking was that since we are limiting the number of submissions this won't have much of an impact.

Overall this rate limit is not scalable. I guess my question is - Is this feature intended for production or is this simply needed to manage cost while Beta testing? If it is intended for production we need to discuss further.

I could write a call back for each request. This way the submissions won't be waiting on one another.

Reply:
For the complexity this introduces, I am not even sure we should go ahead.
But this problem will haunt me. I can find a solid solution for java. But I find it hard to do in NodeJS.